include/eve/pattern.hpp

```diff
@@ -0,0 +1,138 @@
+//==================================================================================================
+/**
+  EVE - Expressive Vector Engine
+  Copyright 2018-2021 Joel FALCOU
+  Copyright 2018-2021 Jean-Thierry LAPRESTE
+
+  Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+  SPDX-License-Identifier: MIT
+**/
+//==================================================================================================
+#pragma once
+
+#include <utility>
+#include <ostream>
+#include <eve/concept/pattern.hpp>
+
+namespace eve
+{
+  //================================================================================================
+  // Index-based pattern holder
+  //================================================================================================
+  template<std::ptrdiff_t... I> struct pattern_t
+  {
+    static inline constexpr auto size(int) noexcept { return sizeof...(I); }
+
+    static inline constexpr bool has_zeros(int) noexcept
+    {
+      return ((I == -1) || ... || false);
+    }
+
+    // Ensure that a pattern don't try to reference non-existent lanes
+    static constexpr bool validate(std::ptrdiff_t N) noexcept
+    {
+      return (((I == -1) || (I<N)) && ... && true);
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, pattern_t const&)
+    {
+      os << "< ";
+      ((os << I << " "),...);
+      return os << ">";
+    }
+
+    constexpr std::ptrdiff_t operator()(std::ptrdiff_t i, int) const noexcept
+    {
+      constexpr std::ptrdiff_t sz = sizeof...(I);
+      constexpr std::ptrdiff_t data[] = {I...};
+
+      return i<sz ? data[i] : -1;
+    }
+
+    template<std::ptrdiff_t... O>
+    constexpr auto is_similar(pattern_t<I..., O...>) const noexcept requires(sizeof...(I)>0)
+    {
+      return true;
+    }
+
+    template<std::ptrdiff_t... O>
+    constexpr auto is_similar(pattern_t<O...>) const  noexcept { return false; }
+
+    constexpr bool operator>  (std::ptrdiff_t n) const noexcept { return ((I  > n || I == -1) && ...); }
+    constexpr bool operator<  (std::ptrdiff_t n) const noexcept { return ((I  < n || I == -1) && ...); }
+    constexpr bool operator>= (std::ptrdiff_t n) const noexcept { return ((I >= n || I == -1) && ...); }
+    constexpr bool operator<= (std::ptrdiff_t n) const noexcept { return ((I <= n || I == -1) && ...); }
+
```

Comment:
Please don't make this operator overloads.

Reply:
Sure